host/s3.go

```diff
@@ -0,0 +1,412 @@
+			)
+			return nil
+		}
+	}
+	return fmt.Errorf("Couldn't find bucket %s among existing ones in the s3 instance", s3ds.config.Bucket)
+}
+
+// removeItemFromS3 removes the item from the s3 instance (for testing purposes)
+func (s3ds *S3DataStream) removeItemFromS3(ctx context.Context, s3Key string) error {
+	slog.Debug(
+		"launching S3 periodic flusher",
+		slog.Attr{Key: "s3key", Value: slog.StringValue(s3Key)},
+	)
+	opCtx, cancel := context.WithTimeout(ctx, S3OpTimeout)
+	defer cancel()
+
+	s3ds.client.DeleteObject(opCtx, &s3.DeleteObjectInput{
+		Bucket: aws.String(s3ds.config.Bucket),
+		Key:    aws.String(s3Key),
+	})
+
+	return nil
+}
+
+// spawnPeriodicFlusher is a S3DataStream method that will create a background routine
+// to flush the batched traces every given interval
+func (s3ds *S3DataStream) spawnPeriodicFlusher(ctx context.Context, interval time.Duration) {
+	go func() {
+		slog.Debug("launching S3 periodic flusher", slog.Attr{Key: "flush-interval", Value: slog.StringValue(interval.String())})
+		flushTicker := time.NewTicker(interval)
+		for {
+			select {
+			case <-ctx.Done():
+				slog.Debug("context died, closing the s3 trace-periodic-flusher")
+				// use new context to submit whatever is left
+				s3ds.submitRecords(context.TODO())
+				return
+			case <-flushTicker.C:
+
+			}
+		}
+	}()
+}
+
+// S3DSConfig belongs to the configuration needed to stablish a connection with an s3 instance
+type S3DSConfig struct {
+	ByteLimit       int64
+	AccessKeyID     string
+	SecretAccessKey string
+	Region          string
+	Endpoint        string
+	Bucket          string
+	Prefix          string
+	FlushInterval   time.Duration
+}
+
+// IsValid checks whether the current configuration is valid or not
+// returns true if the S3 Region, the Bucket, the AccessKeyID and the Secret are set
+// returns false otherwise
+func (s3cfg *S3DSConfig) IsValid() bool {
+	return len(s3cfg.Bucket) > 0 &&
+		len(s3cfg.AccessKeyID) > 0 &&
+		len(s3cfg.SecretAccessKey) > 0 &&
+		len(s3cfg.Region) > 0 &&
+		s3cfg.FlushInterval.Nanoseconds() > 0
+}
+
+// ToAWSconfig makes a quick translation from the given user args
+// into the aws.Config struct -> ready to create the S3 client
+func (s3cfg S3DSConfig) ToAWSconfig() (*aws.Config, error) {
+	if !s3cfg.IsValid() {
+		return nil, fmt.Errorf("non valid s3 configuration, %+v", s3cfg)
+	}
+	cfg, err := config.LoadDefaultConfig(
+		context.TODO(),
+		config.WithRegion(s3cfg.Region),
+		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
+			s3cfg.AccessKeyID,
+			s3cfg.SecretAccessKey,
+			"", // empty session for now
+		)),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &cfg, nil
+}
+
+// TraceBatcher is an internal threathsafe buffer for EventTraces
+type traceBatcher struct {
+	sync.RWMutex
+	limit    int64 // bytes
+	traces   []*TraceEvent
+	totBytes int64
+}
+
+// NewTraceBatcher creates a new empty TraceBatcher
+func newTraceBatcher(limit int64) (*traceBatcher, error) {
```

Comment:
I'd rename this because I thought this is the limit of traces (you documented it below that's the byte size limit but perhaps make it more explicit here)
```suggestion
func newTraceBatcher(byteLimit int64) (*traceBatcher, error) {
```